// @wdio/devtools-service/dist/index.js
import { C as e, S as t, _ as n, a as r, b as i, c as a, d as o, f as s, g as c, h as l, i as u, l as d, m as ee, n as f, o as p, p as te, r as m, s as ne, t as re, u as h, v as ie, x as ae, y as oe } from "./launcher-WmbTdxe3.js";
import { TraceType as g } from "./types.js";
import _ from "node:fs/promises";
import v from "node:path";
import y from "@wdio/logger";
import { SevereServiceError as b } from "webdriverio";
import x from "node:url";
import { WebSocket as S } from "ws";
import { parse as C } from "stack-trace";
import { resolve as se } from "import-meta-resolve";
import ce from "@wdio/reporter";
import w from "fs";
import { createRequire as T } from "node:module";
import { parse as E } from "@babel/parser";
import { existsSync as le, readFileSync as ue } from "node:fs";
import de from "node:os";
//#region src/session.ts
var D = y("@wdio/devtools-service:SessionCapturer"), O = (e) => e.replace(f, ""), fe = (e) => {
	let t = O(e).toLowerCase();
	for (let { level: e, pattern: n } of d) if (n.test(t)) return e;
	return r.some((e) => t.includes(e.toLowerCase())) ? "error" : "log";
}, k = (e, t, n) => ({
	timestamp: Date.now(),
	type: e,
	args: t,
	source: n
}), A = class {
	#e;
	#t = !1;
	#n = {
		log: console.log,
		info: console.info,
		warn: console.warn,
		error: console.error
	};
	#r = process.stdout.write.bind(process.stdout);
	#i = process.stderr.write.bind(process.stderr);
	#a = !1;
	#o = /* @__PURE__ */ new Map();
	constructor(e = {}) {
		this.commandsLog = [], this.sources = /* @__PURE__ */ new Map(), this.mutations = [], this.traceLogs = [], this.consoleLogs = [], this.networkRequests = [];
		let { port: t, hostname: n } = e;
		n && t && (this.#e = new S(`ws://${n}:${t}/worker`), this.#e.on("error", (e) => D.error(`Couldn't connect to devtools backend: ${e.message}`))), this.#s(), this.#c();
	}
	#s() {
		m.forEach((e) => {
			let t = this.#n[e];
			console[e] = (...n) => {
				let r = k(e, n.map((e) => typeof e == "object" && e ? (() => {
					try {
						return JSON.stringify(e);
					} catch {
						return String(e);
					}
				})() : String(e)), h.TEST);
				this.consoleLogs.push(r), this.sendUpstream("consoleLogs", [r]), this.#a = !0;
				let i = t.apply(console, n);
				return this.#a = !1, i;
			};
		});
	}
	#c() {
		let e = (e) => {
			let t = typeof e == "string" ? e : e.toString();
			t.trim() && t.split("\n").filter((e) => e.trim()).forEach((e) => {
				let t = k(fe(e), [O(e)], h.TERMINAL);
				this.consoleLogs.push(t), this.sendUpstream("consoleLogs", [t]);
			});
		}, t = (t, n) => {
			t.write = ((r, ...i) => {
				let a = n.call(t, r, ...i);
				return r && !this.#a && e(r), a;
			});
		};
		t(process.stdout, this.#r), t(process.stderr, this.#i);
	}
	cleanup() {
		m.forEach((e) => {
			console[e] = this.#n[e];
		}), process.stdout.write = this.#r, process.stderr.write = this.#i;
	}
	get isReportingUpstream() {
		return !!this.#e && this.#e?.readyState === S.OPEN;
	}
	async ensureSourceLoaded(e) {
		if (!e) return;
		let t = (e.startsWith("file://") ? x.fileURLToPath(e) : e).split(":")[0];
		if (!(!t || this.sources.has(t))) try {
			let e = (await _.readFile(t, "utf-8")).toString();
			this.sources.set(t, e), this.sendUpstream("sources", { [t]: e });
		} catch {}
	}
	async afterCommand(e, t, n, r, i, a) {
		let s = C(/* @__PURE__ */ Error("")).filter((e) => !!e.getFileName()).map((e) => [
			e.getFileName(),
			e.getLineNumber(),
			e.getColumnNumber()
		].join(":")).filter((e) => !e.includes("/node_modules/") && !e.includes("<anonymous>)") && !e.includes("node:internal") && !e.includes("/dist/")).shift() || "", c = s.startsWith("file://") ? x.fileURLToPath(s) : s, l = c.split(":")[0], u = await _.access(l).then(() => !0, () => !1);
		if (s && !this.sources.has(s) && u) {
			let e = await _.readFile(l, "utf-8");
			this.sources.set(l, e.toString()), this.sendUpstream("sources", { [l]: e.toString() });
		}
		let d = {
			command: t,
			args: n,
			result: r,
			error: i,
			timestamp: Date.now(),
			callSource: a ?? c
		};
		try {
			d.screenshot = await e.takeScreenshot();
		} catch (e) {
			D.warn(`failed to capture screenshot: ${e.message}`);
		}
		this.commandsLog.push(d), this.sendUpstream("commands", [d]), o.includes(t) && await this.#l(e);
	}
	async injectScript(e) {
		if (this.#t) {
			D.info("Script already injected, skipping");
			return;
		}
		if (!e.isBidi) throw new b(`Can not set up devtools for session with id "${e.sessionId}" because it doesn't support WebDriver Bidi`);
		this.#t = !0, D.info("Injecting devtools script...");
		let t = await se("@wdio/devtools-script", import.meta.url), n = `async () => { ${(await _.readFile(x.fileURLToPath(t))).toString()} }`;
		await e.scriptAddPreloadScript({ functionDeclaration: n }), D.info("✓ Script injected successfully");
	}
	async #l(e) {
		if (!this.#t) {
			D.warn("Script not injected, skipping trace capture");
			return;
		}
		try {
			if (!await e.execute(() => window.wdioTraceCollector !== void 0)) {
				D.warn("wdioTraceCollector not loaded yet - page loaded before preload script took effect");
				return;
			}
			let { mutations: t, traceLogs: n, consoleLogs: r, networkRequests: i, metadata: a } = await e.execute(() => window.wdioTraceCollector.getTraceData());
			if (this.metadata = a, Array.isArray(t) && (this.mutations.push(...t), this.sendUpstream("mutations", t)), Array.isArray(n) && (this.traceLogs.push(...n), this.sendUpstream("logs", n)), Array.isArray(r)) {
				let e = r;
				e.forEach((e) => e.source = h.BROWSER), this.consoleLogs.push(...e), this.sendUpstream("consoleLogs", e);
			}
			if (Array.isArray(i)) {
				let e = i;
				this.networkRequests.push(...e), this.sendUpstream("networkRequests", e);
			}
			this.sendUpstream("metadata", a), D.info(`✓ Sent metadata upstream, WS state: ${this.#e?.readyState}`);
		} catch (e) {
			D.error(`Failed to capture trace: ${e.message}`);
		}
	}
	handleLogEntryAdded(e) {
		let t = {
			log: "log",
			info: "info",
			warn: "warn",
			error: "error",
			debug: "log",
			trace: "log"
		}[e.method ?? ""] ?? {
			info: "info",
			warn: "warn",
			error: "error",
			debug: "log"
		}[e.level ?? ""] ?? "log", n = Array.isArray(e.args) ? e.args.map((e) => {
			if (e && "value" in e && e.value !== void 0) try {
				return typeof e.value == "string" ? e.value : JSON.stringify(e.value);
			} catch {
				return String(e.value);
			}
			return `[${e?.type ?? "unknown"}]`;
		}) : e.text ? [e.text] : [], r = {
			timestamp: typeof e.timestamp == "number" ? e.timestamp : Date.now(),
			type: t,
			args: n,
			source: h.BROWSER
		};
		this.consoleLogs.push(r), this.sendUpstream("consoleLogs", [r]);
	}
	handleNetworkRequestStarted(e) {
		try {
			let { request: t, timestamp: n } = e, r = t.request, i = {};
			t.headers && t.headers.forEach((e) => {
				let t = typeof e.name == "string" ? e.name.toLowerCase() : "", n = typeof e.value == "string" ? e.value : typeof e.value == "object" && e.value?.value ? e.value.value : "";
				t && (i[t] = n);
			}), this.#o.set(r, {
				url: t.url,
				method: t.method,
				timestamp: n,
				startTime: performance.now(),
				requestHeaders: i
			});
		} catch (e) {
			D.error(`handleNetworkRequestStarted error: ${e}`);
		}
	}
	handleNetworkResponseCompleted(e) {
		try {
			let { request: t, response: n, timestamp: r } = e, i = t.request, a = this.#o.get(i);
			if (!a) return;
			this.#o.delete(i);
			let o = {};
			n.headers && n.headers.forEach((e) => {
				let t = typeof e.name == "string" ? e.name.toLowerCase() : "", n = typeof e.value == "string" ? e.value : typeof e.value == "object" && e.value?.value ? e.value.value : "";
				t && (o[t] = n);
			});
			let s = o["content-type"]?.trim();
			if (!s || s === "-") return;
			let c = performance.now(), l = {
				id: `${r}-${i}`,
				url: a.url,
				method: a.method,
				status: n.status,
				statusText: n.statusText,
				type: this.#u(a.url, s),
				timestamp: a.timestamp,
				startTime: a.startTime,
				endTime: c,
				time: c - a.startTime,
				requestHeaders: a.requestHeaders,
				responseHeaders: o,
				size: n.bytesReceived
			};
			this.networkRequests.push(l), this.sendUpstream("networkRequests", [l]);
		} catch (e) {
			D.error(`handleNetworkResponseCompleted error: ${e}`);
		}
	}
	handleNetworkFetchError(e) {
		let t = e.request.request;
		this.#o.delete(t);
	}
	#u(e, t) {
		let n = e.toLowerCase(), r = t?.toLowerCase() || "";
		return r.includes("text/html") ? "document" : r.includes("text/css") ? "stylesheet" : r.includes("javascript") || r.includes("ecmascript") ? "script" : r.includes("image/") ? "image" : r.includes("font/") || r.includes("woff") ? "font" : r.includes("application/json") ? "fetch" : n.endsWith(".html") || n.endsWith(".htm") ? "document" : n.endsWith(".css") ? "stylesheet" : n.endsWith(".js") || n.endsWith(".mjs") ? "script" : n.match(/\.(png|jpg|jpeg|gif|svg|webp|ico)$/) ? "image" : n.match(/\.(woff|woff2|ttf|eot|otf)$/) ? "font" : "xhr";
	}
	sendUpstream(e, t) {
		!this.#e || this.#e.readyState !== S.OPEN || this.#e.send(JSON.stringify({
			scope: e,
			data: t
		}));
	}
}, j = T(import.meta.url), M = j("@babel/traverse").default, N = /* @__PURE__ */ new Map(), P;
try {
	let e = j("@cucumber/cucumber-expressions");
	P = {
		CucumberExpression: e.CucumberExpression,
		ParameterTypeRegistry: e.ParameterTypeRegistry
	};
} catch {}
var F;
function I(e) {
	F = e;
}
function pe(e) {
	if (!e || e.startsWith("node:")) return !1;
	let t = e.replace(/\\/g, "/");
	return t.includes("/@wdio/expect-webdriverio/") ? !0 : !t.includes("/node_modules/");
}
function L(e) {
	let t = e;
	return t.parent ? L(t.parent) : e;
}
function R(e) {
	if (e) {
		if (e.type === "Identifier") return e.name;
		if (e.type === "MemberExpression") {
			let t = e.object;
			return t && t.type === "Identifier" ? t.name : void 0;
		}
	}
}
function z(n) {
	if (!w.existsSync(n)) return [];
	let r = E(w.readFileSync(n, "utf-8"), {
		sourceType: "module",
		plugins: s,
		errorRecovery: !0,
		allowReturnOutsideFunction: !0
	}), i = [], a = [], o = (e) => !!e && t.includes(e) || e === "Feature", c = (t) => !!t && e.includes(t), l = (e) => {
		if (e) {
			if (e.type === "StringLiteral") return e.value;
			if (e.type === "TemplateLiteral" && e.expressions.length === 0) return e.quasis.map((e) => e.value.cooked).join("");
		}
	};
	return M(r, {
		enter(e) {
			if (!e.isCallExpression()) return;
			let t = e.node.callee, n = R(t);
			if (n) {
				if (o(n)) {
					let t = l(e.node.arguments?.[0]);
					t && (i.push({
						type: "suite",
						name: t,
						titlePath: [...a, t],
						line: e.node.loc?.start.line,
						column: e.node.loc?.start.column
					}), a.push(t));
				} else if (c(n)) {
					let t = l(e.node.arguments?.[0]);
					t && i.push({
						type: "test",
						name: t,
						titlePath: [...a, t],
						line: e.node.loc?.start.line,
						column: e.node.loc?.start.column
					});
				}
			}
		},
		exit(e) {
			if (!e.isCallExpression()) return;
			let t = e.node.callee, n = R(t);
			if (!n || !o(n)) return;
			let r = (() => {
				let t = e.node.arguments?.[0];
				if (t?.type === "StringLiteral") return t.value;
				if (t?.type === "TemplateLiteral" && t.expressions.length === 0) return t.quasis.map((e) => e.value.cooked).join("");
			})();
			r && a[a.length - 1] === r && a.pop();
		}
	}), i;
}
function B() {
	let e = C(/* @__PURE__ */ Error()), t = (t) => {
		let n = e.find((e) => {
			let n = e.getFileName();
			return !!n && !n.includes("node_modules") && t(e);
		});
		return n ? {
			file: n.getFileName(),
			line: n.getLineNumber(),
			column: n.getColumnNumber()
		} : null;
	};
	return t((e) => {
		let t = e.getFileName();
		return i.test(t) || oe.test(t);
	}) || t((e) => l.test(e.getFileName())) || t((e) => p.test(e.getFileName())) || null;
}
function me(e) {
	let t = e;
	for (let e = 0; e < 6; e++) {
		for (let e of c) {
			let n = v.join(t, e);
			if (w.existsSync(n) && w.statSync(n).isDirectory()) return n;
		}
		let e = v.dirname(t);
		if (e === t) break;
		t = e;
	}
}
var V;
function H() {
	if (V && w.existsSync(V)) return V;
	let e = [{
		dir: process.cwd(),
		depth: 0
	}];
	for (; e.length;) {
		let { dir: t, depth: n } = e.shift();
		if (n > 5) continue;
		let r = v.join(t, "features");
		if (w.existsSync(r) && w.statSync(r).isDirectory()) for (let e of c) {
			let t = v.join(r, e);
			if (w.existsSync(t) && w.statSync(t).isDirectory()) return V = t, t;
		}
		for (let r of w.readdirSync(t)) {
			if (r.startsWith(".")) continue;
			let i = v.join(t, r), a;
			try {
				a = w.statSync(i);
			} catch {
				continue;
			}
			a.isDirectory() && !i.includes("node_modules") && e.push({
				dir: i,
				depth: n + 1
			});
		}
	}
}
function U(e) {
	let t = [];
	for (let n of w.readdirSync(e)) {
		let r = v.join(e, n);
		w.statSync(r).isDirectory() ? t.push(...U(r)) : ee.test(n) && t.push(r);
	}
	return t;
}
function he(e) {
	let t = [], r = w.readFileSync(e, "utf-8").split(/\r?\n/);
	for (let i = 0; i < r.length; i++) {
		let a = r[i], o = a.match(n);
		if (o) {
			let n = o[2], r = n.lastIndexOf("/"), a = n.slice(1, r), s = n.slice(r + 1);
			try {
				t.push({
					kind: "regex",
					regex: new RegExp(a, s),
					file: e,
					line: i + 1,
					column: o.index ?? 0
				});
				continue;
			} catch {}
		}
		let s = a.match(ie);
		if (s) {
			let n = s[1], r = s[3];
			t.push({
				kind: "string",
				keyword: n,
				text: r,
				file: e,
				line: i + 1,
				column: s.index ?? 0
			});
		}
	}
	return t;
}
var W = /* @__PURE__ */ new Map();
function ge(e) {
	let t = W.get(e);
	if (t) return t;
	let n = U(e), r = [];
	for (let e of n) {
		let t = 0;
		try {
			M(E(w.readFileSync(e, "utf-8"), {
				sourceType: "module",
				plugins: s,
				errorRecovery: !0
			}), { CallExpression(n) {
				let i = n.node.callee, a;
				if (i?.type === "Identifier") a = i.name;
				else if (i?.type === "MemberExpression") {
					let e = i.property;
					e?.type === "Identifier" && (a = e.name);
				}
				if (!a || !ae.includes(a)) return;
				let o = n.node.arguments?.[0], s = {
					file: e,
					line: n.node.loc?.start.line ?? 1,
					column: n.node.loc?.start.column ?? 0
				};
				if (o?.type === "RegExpLiteral") r.push({
					kind: "regex",
					regex: new RegExp(o.pattern, o.flags ?? ""),
					...s
				}), t++;
				else if (o?.type === "StringLiteral") {
					if (P && o.value.includes("{")) {
						let e = new P.CucumberExpression(o.value, new P.ParameterTypeRegistry());
						r.push({
							kind: "expression",
							expr: e,
							...s
						});
					} else r.push({
						kind: "string",
						keyword: a,
						text: o.value,
						...s
					});
					t++;
				}
			} });
		} catch {}
		if (t === 0) {
			let t = he(e);
			t.length && r.push(...t);
		}
	}
	return W.set(e, r), r;
}
function _e(e, t) {
	let n = t ? v.extname(t) ? v.dirname(t) : t : void 0, r = n ? me(n) : void 0;
	if (r ||= H(), !r) return;
	let i = ge(r), a = String(e ?? "").trim(), o = a.replace(/^(Given|When|Then|And|But)\s+/i, "").trim(), s = i.find((e) => e.kind === "string" && (o.localeCompare(e.text, "en", { sensitivity: "base" }) === 0 || a.localeCompare(`${e.keyword} ${e.text}`, "en", { sensitivity: "base" }) === 0));
	if (s) return {
		file: s.file,
		line: s.line,
		column: s.column
	};
	let c = i.find((e) => e.kind === "expression" && (() => {
		try {
			return !!e.expr.match(o) || !!e.expr.match(a);
		} catch {
			return !1;
		}
	})());
	if (c) return {
		file: c.file,
		line: c.line,
		column: c.column
	};
	let l = i.find((e) => e.kind === "regex" && (e.regex.test(o) || e.regex.test(a)));
	if (l) return {
		file: l.file,
		line: l.line,
		column: l.column
	};
}
function ve(e) {
	return String(e || "").replace(/^\d+:\s*/, "").replace(/\s+/g, " ").trim();
}
function G(e) {
	return e.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function K(e, t) {
	let n = 1, r = 1;
	for (let i = 0; i < t && i < e.length; i++) e.charCodeAt(i) === 10 ? (n++, r = 1) : r++;
	return {
		line: n,
		column: r
	};
}
function ye(t, n) {
	try {
		let r = w.readFileSync(t, "utf-8"), i = `(['"\`])${G(n)}\\1`, a = String.raw`\b(?:${e.join("|")})\s*\(\s*${i}`, o = new RegExp(a).exec(r);
		if (o && typeof o.index == "number") {
			let { line: e, column: n } = K(r, o.index);
			return {
				file: t,
				line: e,
				column: n
			};
		}
	} catch {}
}
function be(e, n) {
	try {
		let r = w.readFileSync(e, "utf-8"), i = `(['"\`])${G(n)}\\1`, a = String.raw`\b(?:${t.join("|")})\s*\(\s*${i}`, o = new RegExp(a).exec(r);
		if (o && typeof o.index == "number") {
			let { line: t, column: n } = K(r, o.index);
			return {
				file: e,
				line: t,
				column: n
			};
		}
	} catch {}
}
function xe(e, t) {
	let n = String(e?.title ?? "").trim(), r = ve(e?.fullTitle), i = (Array.isArray(e.specs) ? e.specs[0] : void 0) || e.file || e.specFile || t || F;
	if (/^(Given|When|Then|And|But)\b/i.test(n)) {
		let t = _e(n, p.test(String(i)) ? i : void 0);
		if (t) {
			Object.assign(e, t);
			return;
		}
	}
	let a = e.file || (Array.isArray(e.specs) ? e.specs[0] : void 0) || e.specFile || t || F;
	if (a && !p.test(a)) {
		if (!N.has(a)) try {
			N.set(a, z(a));
		} catch {}
		let t = N.get(a);
		if (t?.length) {
			let i = t.find((e) => e.type === "test" && e.name === n && r.includes(e.titlePath.join(" "))) || t.find((e) => e.type === "test" && e.name === n);
			if (i) {
				Object.assign(e, {
					file: a,
					line: i.line,
					column: i.column
				});
				return;
			}
		}
		let i = ye(a, n);
		if (i) {
			Object.assign(e, i);
			return;
		}
	}
	let o = B();
	o && Object.assign(e, o);
}
function Se(e, t, n = []) {
	let r = String(e?.title ?? "").trim(), i = e.file || t || F;
	if (!r || !i) return;
	if (p.test(i)) {
		try {
			let t = w.readFileSync(i, "utf-8").split(/\r?\n/), n = (e) => e.trim().replace(/\s+/g, " "), a = n(r);
			for (let r = 0; r < t.length; r++) {
				let o = t[r].match(ne);
				if (o && n(o[2]) === a) {
					Object.assign(e, {
						file: i,
						line: r + 1,
						column: 1
					});
					return;
				}
			}
		} catch {}
		return;
	}
	try {
		N.has(i) || N.set(i, z(i));
		let t = N.get(i);
		if (t?.length) {
			let a = t.find((e) => e.type === "suite" && Array.isArray(e.titlePath) && e.titlePath.length === n.length && e.titlePath.every((e, t) => e === n[t])) || t.find((e) => e.type === "suite" && e.titlePath.at(-1) === r);
			if (a?.line) {
				Object.assign(e, {
					file: i,
					line: a.line,
					column: a.column
				});
				return;
			}
		}
	} catch {}
	let a = be(i, r);
	a && Object.assign(e, a);
}
//#endregion
//#region src/reporter.ts
var q = /* @__PURE__ */ new Map();
function J(e) {
	let t = e;
	if (t.type === "scenario" && /^\d+$/.test(t.uid)) {
		let n = [
			e.title,
			t.file || "",
			t.parent || "",
			t.cid || "",
			`example-${t.uid}`
		].join("::").split("").reduce((e, t) => (e << 5) - e + t.charCodeAt(0) | 0, 0);
		return `stable-${Math.abs(n).toString(36)}`;
	}
	let n = [t.file || "", String(t.fullTitle || e.title)], r = n.join("::"), i = q.get(r) || 0;
	q.set(r, i + 1), i > 0 && n.push(String(i));
	let a = n.join("::").split("").reduce((e, t) => (e << 5) - e + t.charCodeAt(0) | 0, 0);
	return `stable-${Math.abs(a).toString(36)}`;
}
function Ce() {
	q.clear();
}
function we(e, t) {
	try {
		if (!le(e)) return null;
		let n = ue(e, "utf-8").split("\n"), r = -1;
		for (let e = 0; e < n.length; e++) {
			let i = n[e].trim();
			if ((i.startsWith("Scenario Outline:") || i.startsWith("Scenario:")) && i.includes(t)) {
				r = e;
				break;
			}
		}
		if (r === -1) return null;
		let i = -1;
		for (let e = r; e < n.length; e++) if (n[e].trim().startsWith("Examples:")) {
			i = e;
			break;
		}
		if (i === -1) return null;
		let a = /* @__PURE__ */ new Map(), o = 0, s = !1;
		for (let e = i + 1; e < n.length; e++) {
			let t = n[e].trim();
			if (t.startsWith("Scenario") || t.startsWith("Feature:") || !t && o > 0) break;
			t.startsWith("|") && (s ? (a.set(o, e + 1), o++) : s = !0);
		}
		return a.size > 0 ? a : null;
	} catch (e) {
		return console.error("[Reporter] Failed to parse feature file:", e), null;
	}
}
var Te = class extends ce {
	#e;
	#t;
	#n;
	#r = [];
	constructor(e, t, n = () => {}) {
		super(e), this.#e = t, this.#t = n, Ce();
	}
	onSuiteStart(e) {
		super.onSuiteStart(e);
		let t = e;
		if (t.type === "scenario" && e.file?.endsWith(".feature")) {
			let n = parseInt(t.uid, 10);
			if (!isNaN(n)) {
				let r = we(e.file, e.title);
				if (r?.has(n)) {
					let i = r.get(n);
					t.featureFile = e.file, t.featureLine = i;
				}
			}
		}
		e.uid = J(e), this.#n = e.file, I(e.file), e.title && this.#r.push(e.title), Se(e, this.#n, this.#r), e.file && e.line !== null && (e.callSource = `${e.file}:${e.line}`, this.#t(e.file)), this.#i();
	}
	onTestStart(e) {
		super.onTestStart(e);
		let t = e;
		t.argument?.uri && typeof t.argument?.line == "number" && (t.featureFile = t.argument.uri, t.featureLine = t.argument.line), xe(e, this.#n), e.file && e.line !== null && (e.callSource = `${e.file}:${e.line}`, this.#t(e.file)), e.uid = J(e), this.#i();
	}
	onTestEnd(e) {
		super.onTestEnd(e), this.#i();
	}
	onSuiteEnd(e) {
		super.onSuiteEnd(e), e.title && this.#r[this.#r.length - 1] === e.title && this.#r.pop(), this.#r.length === 0 && (this.#n = void 0, I(void 0)), this.#i();
	}
	#i() {
		if (!this.suites) return;
		let e = [];
		for (let t of Object.values(this.suites)) if (t) {
			let n = t.uid;
			e.push({ [n]: t });
		}
		e.length > 0 && this.#e(e);
	}
	get report() {
		return this.suites;
	}
}, Y = y("@wdio/devtools-service:ScreencastRecorder"), X = class {
	#e = [];
	#t = void 0;
	#n = void 0;
	#r = !1;
	#i;
	#a = 0;
	#o = !1;
	constructor(e = {}) {
		this.#i = {
			...te,
			...e
		};
	}
	async start(e) {
		await this.#s(e) || await this.#l(e);
	}
	async stop() {
		this.#r &&= (this.#t ? await this.#c() : this.#n !== void 0 && this.#u(), !1);
	}
	setStartMarker() {
		this.#o || (this.#o = !0, this.#a = this.#e.length);
	}
	get frames() {
		return this.#e.slice(this.#a);
	}
	get duration() {
		let e = this.frames;
		return e.length < 2 ? 0 : e[e.length - 1].timestamp - e[0].timestamp;
	}
	get isRecording() {
		return this.#r;
	}
	async #s(e) {
		try {
			let t = await (await e.getPuppeteer()).pages();
			return t.length ? (this.#t = await t[0].createCDPSession(), await this.#t.send("Page.startScreencast", {
				format: this.#i.captureFormat,
				quality: this.#i.quality,
				maxWidth: this.#i.maxWidth,
				maxHeight: this.#i.maxHeight
			}), this.#t.on("Page.screencastFrame", async (e) => {
				this.#e.push({
					data: e.data,
					timestamp: Math.round(e.metadata.timestamp * 1e3)
				});
				try {
					await this.#t.send("Page.screencastFrameAck", { sessionId: e.sessionId });
				} catch (e) {
					Y.warn(`Screencast: failed to ack frame — ${e.message}`);
				}
			}), this.#r = !0, Y.info("✓ Screencast recording started (CDP mode)"), !0) : !1;
		} catch {
			return !1;
		}
	}
	async #c() {
		try {
			await this.#t.send("Page.stopScreencast"), Y.info(`✓ Screencast stopped — ${this.#e.length} frame(s) collected`);
		} catch (e) {
			let t = e.message ?? "";
			t.includes("Session closed") || t.includes("Target closed") ? Y.debug("Screencast: CDP session already closed (expected during teardown)") : Y.warn(`Screencast: error stopping CDP — ${t}`);
		} finally {
			this.#t = void 0;
		}
	}
	async #l(e) {
		try {
			let t = await e.takeScreenshot();
			this.#e.push({
				data: t,
				timestamp: Date.now()
			});
			let n = this.#i.pollIntervalMs;
			this.#n = setInterval(async () => {
				try {
					let t = await e.takeScreenshot();
					this.#e.push({
						data: t,
						timestamp: Date.now()
					});
				} catch {
					this.#u();
				}
			}, n), this.#r = !0, Y.info(`✓ Screencast recording started (polling mode, ${n} ms interval)`);
		} catch (e) {
			Y.warn(`Screencast unavailable (${e.message}). Recording will be skipped.`);
		}
	}
	#u() {
		this.#n !== void 0 && (clearInterval(this.#n), this.#n = void 0, Y.info(`✓ Screencast stopped — ${this.#e.length} frame(s) collected`));
	}
}, Ee = T(import.meta.url), Z = y("@wdio/devtools-service:VideoEncoder");
async function De(e, t, n = {}) {
	if (e.length === 0) throw Error("VideoEncoder: no frames to encode");
	let r;
	try {
		r = Ee("fluent-ffmpeg");
	} catch {
		throw Error("VideoEncoder: fluent-ffmpeg is required for screencast encoding. Install it with: npm install fluent-ffmpeg");
	}
	let i = n.captureFormat === "png" ? "png" : "jpg", a = await _.mkdtemp(v.join(de.tmpdir(), "wdio-screencast-"));
	try {
		let n = ["ffconcat version 1.0"];
		for (let t = 0; t < e.length; t++) {
			let r = `frame-${String(t).padStart(6, "0")}.${i}`, o = v.join(a, r);
			await _.writeFile(o, Buffer.from(e[t].data, "base64"));
			let s = e[t + 1]?.timestamp ?? e[t].timestamp + 100, c = Math.max((s - e[t].timestamp) / 1e3, .01);
			n.push(`file '${o}'`), n.push(`duration ${c.toFixed(6)}`);
		}
		let o = v.join(a, `frame-${String(e.length - 1).padStart(6, "0")}.${i}`);
		n.push(`file '${o}'`);
		let s = v.join(a, "manifest.txt");
		await _.writeFile(s, n.join("\n")), Z.info(`VideoEncoder: encoding ${e.length} frames → ${t}`), await new Promise((e, n) => {
			r().input(s).inputOptions([
				"-f",
				"concat",
				"-safe",
				"0"
			]).videoCodec("libvpx").outputOptions([
				"-b:v",
				"1M",
				"-pix_fmt",
				"yuv420p",
				"-vsync",
				"vfr",
				"-auto-alt-ref",
				"0",
				"-disposition:v",
				"default"
			]).output(t).on("end", () => e()).on("error", (e) => {
				let t = e.message || "";
				t.includes("Cannot find ffmpeg") || t.includes("ENOENT") || t.includes("spawn") || t.includes("not found") ? n(/* @__PURE__ */ Error("VideoEncoder: ffmpeg binary not found on PATH. Install ffmpeg: https://ffmpeg.org/download.html")) : n(/* @__PURE__ */ Error(`VideoEncoder: ffmpeg error — ${t}`));
			}).run();
		}), Z.info(`✓ Screencast video saved: ${t}`);
	} finally {
		await _.rm(a, {
			recursive: !0,
			force: !0
		}).catch((e) => {
			Z.warn(`VideoEncoder: failed to clean temp dir — ${e.message}`);
		});
	}
}
//#endregion
//#region src/index.ts
var Oe = re, Q = y("@wdio/devtools-service");
function ke(e) {
	let t = !1, n = new $();
	return n.captureType = g.Standalone, n.beforeSession(e, e), e.beforeCommand = Array.isArray(e.beforeCommand) ? e.beforeCommand : e.beforeCommand ? [e.beforeCommand] : [], e.beforeCommand.push(async function(e) {
		t || (t = !0, n.before(this.capabilities, [], this)), e === "deleteSession" && await n.after();
	}, n.beforeCommand.bind(n)), e.afterCommand = Array.isArray(e.afterCommand) ? e.afterCommand : e.afterCommand ? [e.afterCommand] : [], e.afterCommand.push(n.afterCommand.bind(n)), e;
}
function Ae() {
	let e = process.env.DEVTOOLS_WDIO_CONFIG;
	if (e) return v.isAbsolute(e) ? e : v.resolve(process.cwd(), e);
	let t = process.argv;
	for (let e = 0; e < t.length - 1; e++) if (t[e] === "--config" || t[e] === "-c") {
		let n = t[e + 1];
		if (n && /\.(conf|config)\.(ts|js|cjs|mjs)$/i.test(n)) return v.isAbsolute(n) ? n : v.resolve(process.cwd(), n);
	}
	let n = t.find((e) => /\.conf\.(ts|js|cjs|mjs)$/i.test(e));
	if (n) return v.isAbsolute(n) ? n : v.resolve(process.cwd(), n);
}
var $ = class {
	#e = [];
	#t = new A();
	#n;
	#r = !1;
	#i;
	#a;
	constructor(e = {}) {
		this.captureType = g.Testrunner, this.#a = e.screencast;
	}
	#o = [];
	#s = null;
	#c = !1;
	async before(e, t, n) {
		this.#n = n, this.#t = new A(e["wdio:devtoolsOptions"]);
		try {
			await this.#d(n);
		} catch (e) {
			Q.error(`Failed to inject script at session start: ${e.message}`);
		}
		this.#a?.enabled && (this.#i = new X(this.#a), await this.#i.start(n)), n.execute(() => window.visualViewport).then((e) => this.#t.sendUpstream("metadata", {
			viewport: e || void 0,
			type: this.captureType,
			options: n.options,
			capabilities: n.capabilities
		}));
	}
	beforeSession(e, t) {
		if (!("browserName" in t) && !("platformName" in t)) throw new b("The DevTools hook does not support multiremote yet");
		let n = Ae();
		if (n && this.#t.sendUpstream("config", { configFile: n }), "reporters" in e) {
			let t = this;
			e.reporters = [...e.reporters || [], class extends Te {
				constructor(e) {
					super(e, (e) => t.#t.sendUpstream("suites", e), (e) => {
						t.#t.ensureSourceLoaded(e);
					}), t.#e.push(this);
				}
			}];
		}
	}
	beforeScenario() {
		this.resetStack();
	}
	beforeTest() {
		this.resetStack();
	}
	resetStack() {
		this.#s = null, this.#o = [];
	}
	async beforeCommand(e, t) {
		if (!this.#n) return;
		if (!this.#r && this.#n.isBidi) {
			this.#r = !0, Q.info("Setting up BiDi network event listeners..."), this.#n.on("network.beforeRequestSent", (e) => {
				this.#t.handleNetworkRequestStarted(e);
			}), this.#n.on("network.responseCompleted", (e) => {
				this.#t.handleNetworkResponseCompleted(e);
			}), this.#n.on("network.fetchError", (e) => {
				Q.info(`>>> BiDi fetchError - keys: ${Object.keys(e).join(", ")}`), this.#t.handleNetworkFetchError(e);
			}), this.#n.on("log.entryAdded", (e) => {
				this.#t.handleLogEntryAdded(e);
			});
			try {
				this.#n.sessionSubscribe?.({ events: ["log.entryAdded"] });
			} catch (e) {
				Q.warn(`Could not subscribe to log.entryAdded: ${e.message}`);
			}
			Q.info("✓ BiDi network + log event listeners registered");
		}
		e === "url" && (this.#i?.setStartMarker(), this.#t.sendUpstream("metadata", { url: t[0] })), Error.stackTraceLimit = 20;
		let n = C(/* @__PURE__ */ Error("")).reverse().find((e) => pe(e.getFileName()));
		if (n && this.#o.length === 0 && !a.includes(e)) {
			let r = n.getFileName() ?? void 0, i = r;
			if (r?.startsWith("file://")) try {
				let e = new URL(r);
				i = decodeURIComponent(e.pathname);
			} catch {
				i = r;
			}
			i?.includes("?") && (i = i.split("?")[0]);
			let a = n.getLineNumber() ?? void 0, o = n.getColumnNumber() ?? void 0, s = i === void 0 ? void 0 : `${i}:${a ?? 0}:${o ?? 0}`, c = JSON.stringify({
				command: e,
				args: t,
				src: s
			});
			this.#s !== c && (this.#o.push({
				command: e,
				callSource: s
			}), this.#s = c);
		}
	}
	afterCommand(e, t, n, r) {
		if (this.#c) return;
		let i = this.#o[this.#o.length - 1];
		if (i?.command === e && (this.#o.pop(), this.#n)) return this.#t.afterCommand(this.#n, e, t, n, r, i.callSource);
		u.includes(e) && this.#f(`context-change:${e}`);
	}
	async after() {
		if (!this.#n) return;
		await this.#u(this.#n.sessionId);
		let e = this.#l, { ...t } = this.#n.options, n = {
			mutations: this.#t.mutations,
			logs: this.#t.traceLogs,
			consoleLogs: this.#t.consoleLogs,
			networkRequests: this.#t.networkRequests,
			metadata: {
				type: this.captureType,
				...this.#t.metadata,
				options: t,
				capabilities: this.#n.capabilities
			},
			commands: this.#t.commandsLog,
			sources: Object.fromEntries(this.#t.sources),
			suites: this.#e.map((e) => e.report)
		}, r = v.join(e, `wdio-trace-${this.#n.sessionId}.json`);
		await _.writeFile(r, JSON.stringify(n)), Q.info(`DevTools trace saved to ${r}`), this.#t.cleanup();
	}
	async onReload(e, t) {
		!this.#a?.enabled || !this.#n || (await this.#u(e), this.#i = new X(this.#a), await this.#i.start(this.#n));
	}
	get #l() {
		let e = this.#n?.options;
		return e?.outputDir || e?.rootDir || process.cwd();
	}
	async #u(e) {
		if (!this.#i || (await this.#i.stop(), this.#i.frames.length < 5)) return;
		let t = this.#l, n = `wdio-video-${e}.webm`, r = v.join(t, n);
		try {
			await De(this.#i.frames, r, { captureFormat: this.#a?.captureFormat });
			let t = {
				sessionId: e,
				videoPath: r,
				videoFile: n,
				frameCount: this.#i.frames.length,
				duration: this.#i.duration
			};
			this.#t.sendUpstream("screencast", t);
		} catch (e) {
			Q.warn(`Screencast encode failed: ${e.message}`);
		}
	}
	async #d(e) {
		if (!e.isBidi) throw new b(`Can not set up devtools for session with id "${e.sessionId}" because it doesn't support WebDriver Bidi`);
		await this.#t.injectScript(L(e)), Q.info("✓ Devtools preload script active");
	}
	async #f(e) {
		if (!(!this.#n || this.#c)) try {
			if (this.#c = !0, await this.#n.execute(() => !!window.__WDIO_DEVTOOLS_MARK)) return;
			await this.#t.injectScript(L(this.#n));
		} catch (t) {
			Q.warn(`[inject] failed (reason=${e}): ${t.message}`);
		} finally {
			this.#c = !1;
		}
	}
};
//#endregion
export { g as TraceType, $ as default, Oe as launcher, ke as setupForDevtools };
