// @wdio/devtools-service/dist/launcher-WmbTdxe3.js
import e from "node:path";
import t from "@wdio/logger";
import { remote as n } from "webdriverio";
import r from "node:http";
import { start as i } from "@wdio/devtools-backend";
//#region src/constants.ts
var a = {
	enabled: !1,
	captureFormat: "jpeg",
	quality: 70,
	maxWidth: 1280,
	maxHeight: 720,
	pollIntervalMs: 200
}, o = [
	"url",
	"navigateTo",
	"elementClick",
	"click"
], s = /\x1b\[[0-9;]*m/g, c = [
	"log",
	"info",
	"warn",
	"error"
], l = [
	{
		level: "trace",
		pattern: /\btrace\b/i
	},
	{
		level: "debug",
		pattern: /\bdebug\b/i
	},
	{
		level: "info",
		pattern: /\binfo\b/i
	},
	{
		level: "warn",
		pattern: /\bwarn(ing)?\b/i
	},
	{
		level: "error",
		pattern: /\berror\b/i
	}
], u = [
	"✗",
	"failed",
	"failure"
], d = {
	BROWSER: "browser",
	TEST: "test",
	TERMINAL: "terminal"
}, f = {
	browserName: "chrome",
	"goog:chromeOptions": { args: ["--window-size=1600,1200"] }
}, p = [
	"emit",
	"browsingContextLocateNodes",
	"browsingContextNavigate",
	"waitUntil",
	"getTitle",
	"getUrl",
	"getWindowSize",
	"setWindowSize",
	"deleteSession",
	"findElementFromShadowRoot",
	"findElementsFromShadowRoot",
	"waitForExist",
	"browsingContextGetTree",
	"scriptCallFunction",
	"getElement",
	"execute",
	"findElement",
	"getElementText",
	"getElementShadowRoot",
	"scriptAddPreloadScript",
	"getPuppeteer",
	"takeScreenshot"
], m = [
	"url",
	"back",
	"forward",
	"refresh",
	"switchFrame",
	"newWindow",
	"createWindow",
	"closeWindow"
], h = [
	"typescript",
	"jsx",
	"decorators-legacy",
	"classProperties",
	"dynamicImport"
], g = [
	"it",
	"test",
	"specify",
	"fit",
	"xit"
], _ = [
	"describe",
	"context",
	"suite"
], v = [
	"Given",
	"When",
	"Then",
	"And",
	"But",
	"defineStep"
], y = /\.(?:steps?)\.[cm]?[jt]sx?$/i, b = /(?:^|\/)(?:step[-_]?definitions|steps)\/.+\.[cm]?[jt]sx?$/i, x = /\.(?:test|spec)\.[cm]?[jt]sx?$/i, S = /\.feature$/i, C = /\.(?:[cm]?js|[cm]?ts)x?$/, w = /^\s*(Feature|Scenario(?: Outline)?):\s*(.+)\s*$/i, T = /\b(Given|When|Then|And|But)\s*\(\s*(\/(?:\\.|[^/\\])+\/[gimsuy]*)/, E = /\b(Given|When|Then|And|But)\s*\(\s*(['`])([^'`\\]*(?:\\.[^'`\\]*)*)\2/, D = [
	"step-definitions",
	"step_definitions",
	"steps"
], O = t("@wdio/devtools-service:Launcher");
if (process.env.DEVTOOLS_APP_REUSE === "1") {
	let e = r.Server.prototype.listen;
	r.Server.prototype.listen = function(...t) {
		let n = this.emit.bind(this);
		return this.emit = (e, ...r) => e === "error" && r[0]?.code === "EADDRINUSE" ? (O.warn(`Suppressed EADDRINUSE on rerun (port ${String(t[0])}); reusing existing server from the originating CLI process`), this.emit = n, process.nextTick(() => n("listening")), !0) : n(e, ...r), e.apply(this, t);
	};
}
function k() {
	let t = process.argv;
	for (let n = 0; n < t.length - 1; n++) if (t[n] === "--config" || t[n] === "-c") {
		let r = t[n + 1];
		if (r && /\.(conf|config)\.(ts|js|cjs|mjs)$/i.test(r)) return e.isAbsolute(r) ? r : e.resolve(process.cwd(), r);
	}
	let n = t.find((e) => /\.conf\.(ts|js|cjs|mjs)$/i.test(e));
	if (n) return e.isAbsolute(n) ? n : e.resolve(process.cwd(), n);
}
function A() {
	let t = process.argv, n = [];
	for (let r = 0; r < t.length - 1; r++) if (t[r] === "--spec" || t[r] === "-s") {
		let i = t[r + 1];
		if (!i) continue;
		for (let t of i.split(",")) {
			let r = t.trim();
			r && n.push(e.isAbsolute(r) ? r : e.resolve(process.cwd(), r));
		}
	}
	return n;
}
var j = class {
	#e;
	#t;
	constructor(e) {
		this.#e = e;
	}
	async onPrepare(t, r) {
		try {
			let t = k();
			if (t && !process.env.DEVTOOLS_WDIO_CONFIG && (process.env.DEVTOOLS_WDIO_CONFIG = t, O.info(`Detected config for reruns: ${t}`)), !process.env.DEVTOOLS_WDIO_INITIAL_SPECS) {
				let t = A();
				t.length && (process.env.DEVTOOLS_WDIO_INITIAL_SPECS = t.join(e.delimiter), O.info(`Detected initial specs for Run All: ${t.join(", ")}`));
			}
			let a = process.env.DEVTOOLS_APP_PORT, o = process.env.DEVTOOLS_APP_HOST || this.#e.hostname || "localhost";
			if (process.env.DEVTOOLS_APP_REUSE === "1" && a) {
				O.info(`Reusing existing DevTools app at http://${o}:${a}`), this.#n(r, {
					port: Number(a),
					hostname: o
				});
				return;
			}
			let { port: s } = await i({
				port: this.#e.port,
				hostname: this.#e.hostname
			});
			if (!s) return console.log(`Failed to start server on port ${s}`);
			this.#n(r, {
				port: s,
				hostname: this.#e.hostname || "localhost"
			}), this.#t = await n({
				automationProtocol: "devtools",
				capabilities: {
					...f,
					...this.#e.devtoolsCapabilities
				}
			}), await this.#t.url(`http://localhost:${s}`);
		} catch (e) {
			console.error(e);
		}
	}
	async onComplete() {
		if (this.#t) {
			for (t.setLevel("devtools", "warn"), O.info("Please close the browser window to finish...");;) try {
				await this.#t.getTitle(), await new Promise((e) => setTimeout(e, 1e3));
			} catch {
				O.info("Browser window closed, stopping DevTools app");
				break;
			}
			try {
				await this.#t.deleteSession();
			} catch (e) {
				O.warn("Session already closed or could not be deleted:", e.message);
			}
		}
	}
	#n(e, t) {
		if (Array.isArray(e)) for (let n of e) n["wdio:devtoolsOptions"] = {
			port: t.port,
			hostname: t.hostname || "localhost"
		};
	}
};
//#endregion
export { g as C, _ as S, T as _, u as a, y as b, p as c, o as d, h as f, D as g, x as h, m as i, l, C as m, s as n, S as o, a as p, c as r, w as s, j as t, d as u, E as v, v as x, b as y };
