// @wdio/devtools-backend/dist/runner.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _TestRunner_instances, _TestRunner_child, _TestRunner_lastPayload, _TestRunner_registeredConfigFile, _TestRunner_baseDir, _TestRunner_expectingRerunChild, _TestRunner_resolveGenericCommand, _TestRunner_parseGenericCommand, _TestRunner_buildFilters, _TestRunner_resolveLineNumber, _TestRunner_normaliseSpecFile, _TestRunner_getSpecCandidate, _TestRunner_extractSpecFromCallSource, _TestRunner_toFsPath, _TestRunner_resolveConfigPath, _TestRunner_findConfigFromSpec, _TestRunner_expandDefaultConfigsFor, _TestRunner_dedupeCandidates;
import { spawn } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import url from 'node:url';
import { createRequire } from 'node:module';
import kill from 'tree-kill';
import { parse as shellParse } from 'shell-quote';
import { WDIO_CONFIG_FILENAMES, NIGHTWATCH_CONFIG_FILENAMES } from './types.js';
const require = createRequire(import.meta.url);
const wdioBin = resolveWdioBin();
/**
 * Escape special regex characters in a string
 */
function escapeRegex(str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
// Map (not object) keeps payload-supplied `framework` from reaching
// prototype methods at dispatch time — CodeQL: unvalidated-dynamic-method-call.
const FRAMEWORK_FILTERS = new Map();
FRAMEWORK_FILTERS.set('cucumber', ({ specArg, payload }) => {
    const filters = [];
    // For feature-level suites, run the entire feature file
    if (payload.suiteType === 'feature' && specArg) {
        // Remove any line number from specArg for feature-level execution
        const featureFile = specArg.split(':')[0];
        filters.push('--spec', featureFile);
        return filters;
    }
    // Priority 1: Use feature file with line number for exact scenario targeting (works for examples)
    // Note: Cucumber scenarios are type 'suite', not 'test'
    if (payload.featureFile && payload.featureLine) {
        filters.push('--spec', `${payload.featureFile}:${payload.featureLine}`);
        return filters;
    }
    // Priority 2: For specific test reruns with example row number, use exact regex match
    if (payload.entryType === 'test' && payload.fullTitle) {
        // Cucumber fullTitle format: "1: Scenario name" or "2: Scenario name"
        // Extract the row number and scenario name
        // Avoid ReDoS by removing ambiguous \s* before .* - use string operations instead
        const colonIndex = payload.fullTitle.indexOf(':');
        if (colonIndex > 0) {
            const rowNumber = payload.fullTitle.substring(0, colonIndex);
            const scenarioName = payload.fullTitle.substring(colonIndex + 1).trim();
            // Validate row number is digits only
            if (/^\d+$/.test(rowNumber)) {
                // Use spec file filter
                if (specArg) {
                    filters.push('--spec', specArg);
                }
                // Use regex to match the exact "rowNumber: scenarioName" pattern
                // This ensures we only run that specific example row
                filters.push('--cucumberOpts.name', `^${rowNumber}:\\s*${escapeRegex(scenarioName)}$`);
                return filters;
            }
        }
        // No row number - use plain name filter
        if (specArg) {
            filters.push('--spec', specArg);
        }
        filters.push('--cucumberOpts.name', payload.fullTitle.trim());
        return filters;
    }
    // Suite-level rerun
    if (specArg) {
        filters.push('--spec', specArg);
    }
    return filters;
});
FRAMEWORK_FILTERS.set('mocha', ({ specArg, payload }) => {
    const filters = [];
    if (specArg) {
        filters.push('--spec', specArg);
    }
    // For both tests and suites, use grep to filter
    if (payload.fullTitle) {
        filters.push('--mochaOpts.grep', payload.fullTitle);
    }
    return filters;
});
FRAMEWORK_FILTERS.set('jasmine', ({ specArg, payload }) => {
    const filters = [];
    if (specArg) {
        filters.push('--spec', specArg);
    }
    // For both tests and suites, use grep to filter
    if (payload.fullTitle) {
        filters.push('--jasmineOpts.grep', payload.fullTitle);
    }
    return filters;
});
const DEFAULT_FILTERS = ({ specArg }) => specArg ? ['--spec', specArg] : [];
// Nightwatch CLI: positional spec file + optional --testcase filter
FRAMEWORK_FILTERS.set('nightwatch', ({ specArg, payload }) => {
    const filters = [];
    if (specArg) {
        // Nightwatch doesn't support file:line — strip any trailing line number
        filters.push(specArg.split(':')[0]);
    }
    if (payload.entryType === 'test' && payload.label) {
        filters.push('--testcase', payload.label);
    }
    return filters;
});
// Nightwatch + Cucumber: feature files are resolved via the config's feature_path.
// Never pass .feature files as positional args — Nightwatch rejects them.
// Nightwatch forwards --name and --tags to the underlying Cucumber runner.
FRAMEWORK_FILTERS.set('nightwatch-cucumber', ({ payload }) => {
    const filters = [];
    // Only pass --name for scenario-level reruns. Feature/file-level suites
    // (suiteType === 'feature') run all their scenarios, so no --name filter.
    const isFeatureLevel = payload.suiteType === 'feature' || payload.runAll;
    if (!isFeatureLevel && payload.fullTitle) {
        // Wrap as an anchored exact regex so "Scenario A" never also matches
        // "Scenario A-1" (Cucumber treats --name as a regex).
        const escaped = payload.fullTitle.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        filters.push('--name', `^${escaped}$`);
    }
    return filters;
});
class TestRunner {
    constructor() {
        _TestRunner_instances.add(this);
        _TestRunner_child.set(this, void 0);
        _TestRunner_lastPayload.set(this, void 0);
        _TestRunner_registeredConfigFile.set(this, void 0);
        _TestRunner_baseDir.set(this, process.cwd()
        // Set on rerun spawn; consumed once by the next worker handshake so the
        // accumulated suite tree is preserved instead of wiped.
        );
        // Set on rerun spawn; consumed once by the next worker handshake so the
        // accumulated suite tree is preserved instead of wiped.
        _TestRunner_expectingRerunChild.set(this, false);
    }
    consumeRerunChildFlag() {
        const v = __classPrivateFieldGet(this, _TestRunner_expectingRerunChild, "f");
        __classPrivateFieldSet(this, _TestRunner_expectingRerunChild, false, "f");
        return v;
    }
    registerConfigFile(configFile) {
        if (configFile && fs.existsSync(configFile)) {
            __classPrivateFieldSet(this, _TestRunner_registeredConfigFile, configFile, "f");
        }
    }
    async run(payload) {
        if (__classPrivateFieldGet(this, _TestRunner_child, "f")) {
            this.stop();
            await new Promise((resolve) => setTimeout(resolve, 500));
        }
        // devtoolsHost/Port in the payload = REUSE handshake (rerun child).
        __classPrivateFieldSet(this, _TestRunner_expectingRerunChild, Boolean(payload.devtoolsHost && payload.devtoolsPort), "f");
        const isNightwatch = (payload.framework || '')
            .toLowerCase()
            .startsWith('nightwatch');
        // Used when a plugin supplies its own rerun template (e.g. selenium —
        // runs under mocha/jest/vitest/cucumber, none of which use wdioBin).
        const isGenericShell = !isNightwatch && Boolean(payload.rerunCommand || payload.launchCommand);
        const childEnv = { ...process.env };
        if (payload.devtoolsHost && payload.devtoolsPort) {
            childEnv.DEVTOOLS_APP_HOST = payload.devtoolsHost;
            childEnv.DEVTOOLS_APP_PORT = String(payload.devtoolsPort);
            childEnv.DEVTOOLS_APP_REUSE = '1';
        }
        let child;
        if (isGenericShell) {
            const command = __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_resolveGenericCommand).call(this, payload);
            __classPrivateFieldSet(this, _TestRunner_baseDir, process.env.DEVTOOLS_RUNNER_CWD || process.cwd(), "f");
            const { file, args } = __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_parseGenericCommand).call(this, command);
            child = spawn(file, args, {
                cwd: __classPrivateFieldGet(this, _TestRunner_baseDir, "f"),
                env: childEnv,
                stdio: 'inherit',
                detached: false
            });
        }
        else {
            const configPath = __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_resolveConfigPath).call(this, payload);
            __classPrivateFieldSet(this, _TestRunner_baseDir, process.env.DEVTOOLS_RUNNER_CWD || path.dirname(configPath), "f");
            let args;
            if (isNightwatch) {
                const nightwatchBin = resolveNightwatchBin(__classPrivateFieldGet(this, _TestRunner_baseDir, "f"));
                args = [
                    nightwatchBin,
                    '--config',
                    configPath,
                    ...__classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_buildFilters).call(this, payload)
                ].filter(Boolean);
            }
            else {
                args = [
                    wdioBin,
                    'run',
                    configPath,
                    ...__classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_buildFilters).call(this, payload)
                ].filter(Boolean);
            }
            if (isNightwatch) {
                if (payload.entryType === 'test' && payload.label) {
                    childEnv.DEVTOOLS_RERUN_ENTRY_TYPE = 'test';
                    childEnv.DEVTOOLS_RERUN_LABEL = payload.label;
                }
                else {
                    delete childEnv.DEVTOOLS_RERUN_ENTRY_TYPE;
                    delete childEnv.DEVTOOLS_RERUN_LABEL;
                }
            }
            child = spawn(process.execPath, args, {
                cwd: __classPrivateFieldGet(this, _TestRunner_baseDir, "f"),
                env: childEnv,
                stdio: 'inherit',
                detached: false
            });
        }
        __classPrivateFieldSet(this, _TestRunner_child, child, "f");
        __classPrivateFieldSet(this, _TestRunner_lastPayload, payload, "f");
        child.once('close', () => {
            __classPrivateFieldSet(this, _TestRunner_child, undefined, "f");
            __classPrivateFieldSet(this, _TestRunner_lastPayload, undefined, "f");
            __classPrivateFieldSet(this, _TestRunner_baseDir, process.cwd(), "f");
        });
        await new Promise((resolve, reject) => {
            child.once('spawn', resolve);
            child.once('error', (error) => {
                __classPrivateFieldSet(this, _TestRunner_child, undefined, "f");
                __classPrivateFieldSet(this, _TestRunner_lastPayload, undefined, "f");
                __classPrivateFieldSet(this, _TestRunner_baseDir, process.cwd(), "f");
                __classPrivateFieldSet(this, _TestRunner_expectingRerunChild, false, "f");
                reject(error);
            });
        });
    }
    stop() {
        if (!__classPrivateFieldGet(this, _TestRunner_child, "f") || !__classPrivateFieldGet(this, _TestRunner_child, "f").pid) {
            return;
        }
        const pid = __classPrivateFieldGet(this, _TestRunner_child, "f").pid;
        const child = __classPrivateFieldGet(this, _TestRunner_child, "f");
        // SIGTERM, then SIGKILL after 1s — Jest swallows SIGTERM and keeps
        // running until the current test resolves.
        kill(pid, 'SIGTERM', (err) => {
            if (err) {
                kill(pid, 'SIGKILL');
            }
        });
        const sigkillTimer = setTimeout(() => {
            kill(pid, 'SIGKILL');
        }, 1000);
        // Cancel SIGKILL on clean exit — the PID slot may have been recycled.
        child.once('close', () => clearTimeout(sigkillTimer));
        // Clear immediately so a follow-up /api/tests/run isn't blocked by the
        // stale #child guard before SIGKILL lands.
        __classPrivateFieldSet(this, _TestRunner_child, undefined, "f");
        __classPrivateFieldSet(this, _TestRunner_lastPayload, undefined, "f");
        __classPrivateFieldSet(this, _TestRunner_expectingRerunChild, false, "f");
        __classPrivateFieldSet(this, _TestRunner_baseDir, process.cwd(), "f");
    }
}
_TestRunner_child = new WeakMap(), _TestRunner_lastPayload = new WeakMap(), _TestRunner_registeredConfigFile = new WeakMap(), _TestRunner_baseDir = new WeakMap(), _TestRunner_expectingRerunChild = new WeakMap(), _TestRunner_instances = new WeakSet(), _TestRunner_resolveGenericCommand = function _TestRunner_resolveGenericCommand(payload) {
    const template = payload.rerunCommand;
    const fallback = payload.launchCommand || '';
    const isTargetedRerun = !payload.runAll &&
        (payload.entryType === 'test' || payload.entryType === 'suite') &&
        Boolean(payload.label || payload.fullTitle);
    if (template && isTargetedRerun) {
        const name = payload.label || payload.fullTitle || '';
        return template.replace(/\{\{testName\}\}/g, name);
    }
    return fallback || template || '';
}, _TestRunner_parseGenericCommand = function _TestRunner_parseGenericCommand(command) {
    const tokens = shellParse(command).filter((token) => typeof token === 'string');
    if (tokens.length === 0) {
        throw new Error('Invalid generic command: empty command');
    }
    const [file, ...args] = tokens;
    return { file, args };
}, _TestRunner_buildFilters = function _TestRunner_buildFilters(payload) {
    const framework = (payload.framework || '').toLowerCase();
    const specFile = payload.runAll
        ? undefined
        : __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_normaliseSpecFile).call(this, payload);
    const line = specFile ? __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_resolveLineNumber).call(this, payload) : undefined;
    const specArg = specFile
        ? line
            ? `${specFile}:${line}`
            : specFile
        : undefined;
    const candidateBuilder = FRAMEWORK_FILTERS.get(framework);
    const builder = typeof candidateBuilder === 'function'
        ? candidateBuilder
        : DEFAULT_FILTERS;
    const baseFilters = builder({ specArg, payload });
    // Scope "Run All" to the user's original --spec args. Nightwatch resolves specs via its own filter.
    if (payload.runAll && !framework.startsWith('nightwatch')) {
        const initialSpecs = process.env.DEVTOOLS_WDIO_INITIAL_SPECS;
        if (initialSpecs) {
            const specs = initialSpecs.split(path.delimiter).filter(Boolean);
            for (const spec of specs) {
                baseFilters.push('--spec', spec);
            }
        }
    }
    return baseFilters;
}, _TestRunner_resolveLineNumber = function _TestRunner_resolveLineNumber(payload) {
    if (payload.lineNumber && payload.lineNumber > 0) {
        return payload.lineNumber;
    }
    const source = payload.callSource?.trim() || __classPrivateFieldGet(this, _TestRunner_lastPayload, "f")?.callSource;
    if (!source) {
        return __classPrivateFieldGet(this, _TestRunner_lastPayload, "f")?.lineNumber;
    }
    const match = /:(\d+)(?::\d+)?$/.exec(source);
    return match ? Number(match[1]) : __classPrivateFieldGet(this, _TestRunner_lastPayload, "f")?.lineNumber;
}, _TestRunner_normaliseSpecFile = function _TestRunner_normaliseSpecFile(payload) {
    const candidate = __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_getSpecCandidate).call(this, payload);
    return candidate ? __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_toFsPath).call(this, candidate) : undefined;
}, _TestRunner_getSpecCandidate = function _TestRunner_getSpecCandidate(payload) {
    return (payload?.specFile ||
        __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_extractSpecFromCallSource).call(this, payload?.callSource) ||
        __classPrivateFieldGet(this, _TestRunner_lastPayload, "f")?.specFile);
}, _TestRunner_extractSpecFromCallSource = function _TestRunner_extractSpecFromCallSource(source) {
    if (!source) {
        return undefined;
    }
    const match = /^(.*?):\d+:\d+$/.exec(source.trim());
    return match?.[1] ?? source;
}, _TestRunner_toFsPath = function _TestRunner_toFsPath(candidate) {
    const filePath = candidate.startsWith('file://')
        ? url.fileURLToPath(candidate)
        : candidate;
    return path.isAbsolute(filePath)
        ? filePath
        : path.resolve(__classPrivateFieldGet(this, _TestRunner_baseDir, "f"), filePath);
}, _TestRunner_resolveConfigPath = function _TestRunner_resolveConfigPath(payload) {
    const specCandidate = __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_getSpecCandidate).call(this, payload);
    const specDir = specCandidate
        ? path.dirname(__classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_toFsPath).call(this, specCandidate))
        : undefined;
    const isNightwatch = (payload?.framework || '')
        .toLowerCase()
        .startsWith('nightwatch');
    const candidates = __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_dedupeCandidates).call(this, [
        payload?.configFile,
        __classPrivateFieldGet(this, _TestRunner_lastPayload, "f")?.configFile,
        __classPrivateFieldGet(this, _TestRunner_registeredConfigFile, "f"),
        process.env.DEVTOOLS_WDIO_CONFIG,
        process.env.DEVTOOLS_NIGHTWATCH_CONFIG,
        __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_findConfigFromSpec).call(this, specCandidate, isNightwatch),
        ...__classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_expandDefaultConfigsFor).call(this, __classPrivateFieldGet(this, _TestRunner_baseDir, "f"), isNightwatch),
        ...__classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_expandDefaultConfigsFor).call(this, path.resolve(__classPrivateFieldGet(this, _TestRunner_baseDir, "f"), 'example'), isNightwatch),
        ...__classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_expandDefaultConfigsFor).call(this, specDir, isNightwatch)
    ]);
    for (const candidate of candidates) {
        const resolved = __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_toFsPath).call(this, candidate);
        if (fs.existsSync(resolved)) {
            return resolved;
        }
    }
    const runner = isNightwatch ? 'Nightwatch' : 'WDIO';
    throw new Error(`Cannot locate ${runner} config. Tried:\n${candidates
        .map((c) => ` • ${__classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_toFsPath).call(this, c)}`)
        .join('\n')}`);
}, _TestRunner_findConfigFromSpec = function _TestRunner_findConfigFromSpec(specFile, nightwatch = false) {
    if (!specFile) {
        return undefined;
    }
    const filenames = nightwatch
        ? [...NIGHTWATCH_CONFIG_FILENAMES, ...WDIO_CONFIG_FILENAMES]
        : WDIO_CONFIG_FILENAMES;
    const fsSpec = __classPrivateFieldGet(this, _TestRunner_instances, "m", _TestRunner_toFsPath).call(this, specFile);
    let dir = path.dirname(fsSpec);
    const root = path.parse(dir).root;
    while (dir && dir !== root) {
        for (const file of filenames) {
            const candidate = path.join(dir, file);
            if (fs.existsSync(candidate)) {
                return candidate;
            }
        }
        const parent = path.dirname(dir);
        if (parent === dir) {
            break;
        }
        dir = parent;
    }
    return undefined;
}, _TestRunner_expandDefaultConfigsFor = function _TestRunner_expandDefaultConfigsFor(baseDir, nightwatch = false) {
    if (!baseDir) {
        return [];
    }
    const filenames = nightwatch
        ? [...NIGHTWATCH_CONFIG_FILENAMES, ...WDIO_CONFIG_FILENAMES]
        : WDIO_CONFIG_FILENAMES;
    return filenames.map((file) => path.resolve(baseDir, file));
}, _TestRunner_dedupeCandidates = function _TestRunner_dedupeCandidates(values) {
    return Array.from(new Set(values.filter((value) => typeof value === 'string' && value.length > 0)));
};
function resolveNightwatchBin(baseDir) {
    const envOverride = process.env.DEVTOOLS_NIGHTWATCH_BIN;
    if (envOverride) {
        const resolved = path.isAbsolute(envOverride)
            ? envOverride
            : path.resolve(process.cwd(), envOverride);
        if (fs.existsSync(resolved)) {
            return resolved;
        }
    }
    // Walk up from baseDir looking for node_modules/nightwatch/package.json
    // and resolve the actual JS entry (avoids running the shell-script wrapper
    // at node_modules/.bin/nightwatch directly via node).
    let dir = baseDir;
    const root = path.parse(dir).root;
    while (dir !== root) {
        const nightwatchPkgPath = path.join(dir, 'node_modules', 'nightwatch', 'package.json');
        if (fs.existsSync(nightwatchPkgPath)) {
            try {
                const pkg = JSON.parse(fs.readFileSync(nightwatchPkgPath, 'utf8'));
                const nightwatchDir = path.join(dir, 'node_modules', 'nightwatch');
                const binEntry = typeof pkg.bin === 'string'
                    ? pkg.bin
                    : (pkg.bin?.nightwatch ?? pkg.bin?.nw);
                if (binEntry) {
                    const jsPath = path.resolve(nightwatchDir, binEntry);
                    if (fs.existsSync(jsPath)) {
                        return jsPath;
                    }
                }
            }
            catch {
                // malformed package.json — continue walking
            }
        }
        const parent = path.dirname(dir);
        if (parent === dir) {
            break;
        }
        dir = parent;
    }
    throw new Error('Cannot find nightwatch binary. Install nightwatch locally or set DEVTOOLS_NIGHTWATCH_BIN env var.');
}
function resolveWdioBin() {
    const envOverride = process.env.DEVTOOLS_WDIO_BIN;
    if (envOverride) {
        const overriddenPath = path.isAbsolute(envOverride)
            ? envOverride
            : path.resolve(process.cwd(), envOverride);
        if (!fs.existsSync(overriddenPath)) {
            throw new Error(`DEVTOOLS_WDIO_BIN "${overriddenPath}" does not exist or is not accessible`);
        }
        return overriddenPath;
    }
    try {
        const cliEntry = require.resolve('@wdio/cli');
        const candidate = path.resolve(path.dirname(cliEntry), '../bin/wdio.js');
        if (!fs.existsSync(candidate)) {
            throw new Error(`Derived WDIO bin "${candidate}" does not exist`);
        }
        return candidate;
    }
    catch (error) {
        throw new Error(`Failed to resolve WDIO binary. Provide DEVTOOLS_WDIO_BIN env var. ${error.message}`);
    }
}
export const testRunner = new TestRunner();
